### PR TITLE
ECF-RP-501: Use notify id for callbacks on nomination emails

### PR DESCRIPTION
**Pull request by ethanmills**
### Context
Per this comment: https://github.com/DFE-Digital/early-careers-framework/pull/248#discussion_r615936569

### Changes proposed in this pull request
- Change nomination emails callback to use notify id
- Record delivered_at for nomination emails

### Testing
Unit tests

### Review Checks
- [x] All pages have automated accessibility checks via cypress
- [x] All `School` queries are correctly scoped - `eligible` when they need to be

### How can I view this in a review app?
N/A

### CODE REVIEW

File: app/controllers/api/notify_callbacks_controller.rb [hunk 1]
```diff
@@ -2,30 +2,23 @@
...
-      email.update!(notify_status: params[:status]) if email
-    else
-      email = PartnershipNotificationEmail.find_by(notify_id: params[:id])
-      email.update!(notify_status: params[:status], delivered_at: params[:sent_at]&.to_datetime) if email
-    end
+    log_email if failed_email?
 
-    log_email if failed_email?(email)
+    email = PartnershipNotificationEmail.find_by(notify_id: params[:id]) || NominationEmail.find_by(notify_id: params[:id])
+    email.update!(notify_status: params[:status], delivered_at: params[:sent_at]&.to_datetime) if email
```
Comment (kicferk1) [inline app/controllers/api/notify_callbacks_controller.rb:10]: `email&.update!(...)`?

File: db/migrate/20210415163231_add_notify_fields_to_nomination_email.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+class AddNotifyFieldsToNominationEmail < ActiveRecord::Migration[6.1]
+  def change
+    change_table :nomination_emails, bulk: true do
+      add_column :nomination_emails, :notify_id, :string
+      add_column :nomination_emails, :delivered_at, :datetime
+    end
+
+    add_index :nomination_emails, :notify_id
+  end
+end
```
Comment (kicferk1) [inline db/migrate/20210415163231_add_notify_fields_to_nomination_email.rb:12]: In a dream world we would do single table inheritance on them, buuut I think that's probably alright.

File: app/controllers/api/notify_callbacks_controller.rb [hunk 3]
```diff
@@ -2,30 +2,23 @@
...
-      email = NominationEmail.find_by(token: params[:reference])
-      email.update!(notify_status: params[:status]) if email
-    else
-      email = PartnershipNotificationEmail.find_by(notify_id: params[:id])
-      email.update!(notify_status: params[:status], delivered_at: params[:sent_at]&.to_datetime) if email
-    end
+    log_email if failed_email?
 
-    log_email if failed_email?(email)
+    email = PartnershipNotificationEmail.find_by(notify_id: params[:id]) || NominationEmail.find_by(notify_id: params[:id])
```
Comment (kicferk1) [inline app/controllers/api/notify_callbacks_controller.rb:9]: Do we also need to update `notify_id` on NominationEmails somewhere that's sending them?

File: db/migrate/20210415163231_add_notify_fields_to_nomination_email.rb (hunk 2 above)
Comment (amazimbe) [reply db/migrate/20210415163231_add_notify_fields_to_nomination_email.rb:12]: Just what I was thinking too but we can do that in future when things have calmed down

File: app/controllers/api/notify_callbacks_controller.rb (hunk 3 above)
Comment (amazimbe) [reply app/controllers/api/notify_callbacks_controller.rb:9]: I think we can remove the reference param in the mailer too

Comment (ethanmills) [reply app/controllers/api/notify_callbacks_controller.rb:9]: Good spot both

**Review by kicferk1 — APPROVED** [review]

**Review by amazimbe — APPROVED** [review]

**Pull request merged**